File: templates/python/decomposition_stl.py
"""decomposition_stl: 时序分解四联图（原始/趋势/季节/残差，居中移动平均法）."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from theme import apply_theme
from palette import cycle

def make_figure(title='Time-series decomposition'):
    apply_theme(fig_size=(7, 6.5))
    rng = np.random.default_rng(9)
    n_years, period = 8, 12
    n = n_years * period
    t = np.arange(n)
    y = 20 + 0.08 * t + 2 * np.sin(2 * np.pi * t / 96) \
        + 4 * np.sin(2 * np.pi * t / period) \
        + 1.5 * np.cos(4 * np.pi * t / period) \
        + rng.normal(0, 0.8, n)
    # 2x12 居中移动平均求趋势
    w = np.ones(period) / period
    ma = np.convolve(y, w, mode='valid')
    trend = np.full(n, np.nan)
    trend[period // 2: period // 2 + len(ma) - 1] = (ma[:-1] + ma[1:]) / 2
    detr = y - trend
    seas_m = np.array([np.nanmean(detr[m::period]) for m in range(period)])
    seas_m -= seas_m.mean()
    seasonal = np.tile(seas_m, n_years)
    resid = y - trend - seasonal
    parts = [(y, 'observed'), (trend, 'trend'),
             (seasonal, 'seasonal'), (resid, 'residual')]
    fig, axes = plt.subplots(4, 1, sharex=True)
    for ax, (v, name), i in zip(axes, parts, range(4)):
        if name == 'residual':
            ax.plot(t, v, '.', color=cycle(3), markersize=3)
            ax.axhline(0, color='#666666', linewidth=0.8)
        else:
            ax.plot(t, v, color=cycle(i), linewidth=1.2)
        ax.set_ylabel(name)
        ax.grid(True, linestyle=':', alpha=0.5)
    axes[0].set_title(title)
    axes[-1].set_xlabel('time (month index)')
    fig.tight_layout(h_pad=0.4)
    return fig

if __name__ == '__main__':
    fig = make_figure(); fig.savefig(__file__.replace('.py', '.png'), dpi=150)

File: templates/matlab/decomposition_stl.m
function fig = decomposition_stl()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme(); rng(9);
    n_years = 8; period = 12; n = n_years*period;
    t = 0:n-1;
    y = 20 + 0.08*t + 2*sin(2*pi*t/96) + 4*sin(2*pi*t/period) ...
        + 1.5*cos(4*pi*t/period) + 0.8*randn(1, n);
    % 2x12 centered moving average for the trend
    w = ones(1, period)/period;
    ma = conv(y, w, 'valid');
    trend = nan(1, n);
    trend(period/2+1 : period/2+numel(ma)-1) = (ma(1:end-1) + ma(2:end))/2;
    detr = y - trend;
    seas_m = zeros(1, period);
    for m = 1:period
        seas_m(m) = mean(detr(m:period:end), 'omitnan');
    end
    seas_m = seas_m - mean(seas_m);
    seasonal = repmat(seas_m, 1, n_years);
    resid = y - trend - seasonal;
    parts = {y, trend, seasonal, resid};
    names = {'observed', 'trend', 'seasonal', 'residual'};
    fig = figure('Position', [100 100 700 650]);
    for i = 1:4
        ax = subplot(4, 1, i); hold(ax, 'on');
        if i == 4
            plot(t, parts{i}, '.', 'Color', palette('cat', 4), 'MarkerSize', 7);
            yline(0, '-', 'Color', [0.4 0.4 0.4], 'LineWidth', 0.8);
        else
            plot(t, parts{i}, 'Color', palette('cat', i), 'LineWidth', 1.2);
        end
        ylabel(names{i}); grid on; xlim([0 n-1]);
        if i == 1, title('Time-series decomposition'); end
        if i < 4, set(ax, 'XTickLabel', []); end
    end
    xlabel('time (month index)');
end
